subprojects/core/src/main/java/org/gradle/caching/internal/tasks/TarTaskOutputPacker.java

```diff
@@ -127,77 +120,82 @@ private void packMetadata(TaskOutputOriginWriter writeMetadata, TarArchiveOutput
         }
         String propertyPath = "property-" + propertyName;
-        if (!outputFile.exists()) {
+        if (!root.exists()) {
             storeMissingProperty(propertyPath, tarOutput);
             return 1;
         }
         switch (propertySpec.getOutputType()) {
             case DIRECTORY:
-                return 1 + storeDirectoryProperty(propertyPath, outputFile, tarOutput);
+                return storeDirectoryProperty(propertyPath, root, outputSnapshots, tarOutput);
             case FILE:
-                storeFileProperty(propertyPath, outputFile, tarOutput);
+                storeFileProperty(propertyPath, root, tarOutput);
                 return 1;
             default:
                 throw new AssertionError();
         }
     }
 
-    private long storeDirectoryProperty(String propertyPath, File directory, final TarArchiveOutputStream tarOutput) throws IOException {
+    private long storeDirectoryProperty(String propertyPath, File directory, Map<String, FileContentSnapshot> outputSnapshots, final TarArchiveOutputStream tarOutput) throws IOException {
         if (!directory.isDirectory()) {
             throw new IllegalArgumentException(String.format("Expected '%s' to be a directory", directory));
         }
+
+        long entries = 0;
+
         final String propertyRoot = propertyPath + "/";
         createTarEntry(propertyRoot, 0, UnixStat.DIR_FLAG | UnixStat.DEFAULT_DIR_PERM, tarOutput);
         tarOutput.closeArchiveEntry();
+        entries++;
 
-        class CountingFileVisitor implements FileVisitor {
+        String rootAbsolutePath = directory.getAbsolutePath();
+        URI rootUri = directory.toURI();
 
-            private long entries;
-
-            @Override
-            public void visitDir(FileVisitDetails dirDetails) {
-                try {
-                    ++entries;
-                    storeDirectoryEntry(dirDetails, propertyRoot, tarOutput);
-                } catch (IOException e) {
-                    throw new UncheckedIOException(e);
-                }
+        for (Map.Entry<String, FileContentSnapshot> entry : outputSnapshots.entrySet()) {
+            String absolutePath = entry.getKey();
+            // We've already created the directory for the property
+            if (absolutePath.equals(rootAbsolutePath)) {
+                continue;
             }
-
-            @Override
-            public void visitFile(FileVisitDetails fileDetails) {
-                try {
-                    ++entries;
-                    String path = propertyRoot + fileDetails.getRelativePath().getPathString();
-                    storeFileEntry(fileDetails.getFile(), path, fileDetails.getSize(), fileDetails.getMode(), tarOutput);
-                } catch (IOException e) {
-                    throw new UncheckedIOException(e);
-                }
+            File file = new File(absolutePath);
+            String relativePath = rootUri.relativize(file.toURI()).toString();
```

Comment:
We don't capture relative paths for outputs, so I don't think that should be possible.